base_layer/core/src/transactions/transaction_components/key_manager_output_builder.rs

```diff
@@ -150,33 +149,55 @@ impl KeyManagerOutputBuilder {
         &self.covenant
     }
 
-    pub async fn sign_as_sender_and_receiver<KM: BaseLayerKeyManagerInterface>(
-        &mut self,
+    pub async fn sign_as_sender_and_receiver_using_key_id<KM: BaseLayerKeyManagerInterface>(
+        mut self,
         key_manager: &KM,
-        sender_offset_private_key: &PrivateKey,
-    ) -> Result<(), TransactionError> {
+        sender_offset_private_key_id: &KeyId<PublicKey>,
+    ) -> Result<Self, TransactionError> {
         let script = self
             .script
             .as_ref()
             .ok_or_else(|| TransactionError::ValidationError("Cannot sign metadata without a script".to_string()))?;
-        let metadata_signature = TransactionOutput::create_metadata_signature_with_key_id(
-            key_manager,
-            TransactionOutputVersion::get_current_version(),
-            self.value,
-            &self.spending_key_id,
-            script,
+        let sender_offset_public_key = key_manager
+            .get_public_key_at_key_id(sender_offset_private_key_id)
+            .await?;
+        let metadata_message = TransactionOutput::build_metadata_signature_message(
+            &self.version,
+            &script,
             &self.features,
-            sender_offset_private_key,
             &self.covenant,
             &self.encrypted_data,
             self.minimum_value_promise,
-        )
-        .await?;
-        self.sender_offset_public_key = Some(PublicKey::from_secret_key(sender_offset_private_key));
+        );
+        let ephemeral_commitment_nonce_id = key_manager
+            .get_next_key_id(CoreKeyManagerBranch::Nonce.get_branch_key())
+            .await?;
+        let ephemeral_pubkey_nonce_id = key_manager
+            .get_next_key_id(CoreKeyManagerBranch::Nonce.get_branch_key())
+            .await?;
+        let ephemeral_pubkey = key_manager.get_public_key_at_key_id(&ephemeral_pubkey_nonce_id).await?;
+        let ephemeral_commitment = key_manager
+            .get_metadata_signature_ephemeral_commitment(&ephemeral_commitment_nonce_id, self.features.range_proof_type)
+            .await?;
+        let metadata_signature = key_manager
+            .get_metadata_signature(
```

Comment:
Since you are calling `get_metadata_signature` here on the keymanager service, why not just pass in all the inputs and return all the ephemeral keys generated? I say this because there is similar code in the coin base builder and you seem to be asking for data that is used in this call and then thrown away